netCDF4/_netCDF4.pyx

```diff
@@ -4034,16 +4034,16 @@ rename a `netCDF4.Variable` attribute named `oldname` to `newname`."""
                 # for non-string vlen arrays, if data is not multi-dim, or
                 # not an object array, assume it represents a single element
                 # of the vlen var.
-                if not hasattr(data,'ndim') or data.dtype.kind != 'O':
+                if not isinstance(data, numpy.ndarray) or data.dtype.kind != 'O':
                     # issue 458, allow Ellipsis to be used for scalar var
                     if type(elem) == type(Ellipsis) and not\
                        len(self.dimensions): elem = 0
                     self._assign_vlen(elem, data)
                     return
 
-        # A numpy array is needed. Convert if necessary.
-        # assume it's a numpy or masked array if it has an 'ndim' attribute.
-        if not hasattr(data,'ndim'):
+        # A numpy or masked array (or an object supporting the buffer interface) is needed.
+        # Convert if necessary.
+        if not ma.isMA(data) and not (hasattr(data,'data') and isinstance(data.data,buffer)):
```

Comment:
Check that the object is a masked array, or has a ``data`` attribute that is of type ``buffer``.